### ISSUE TITLE
Integration with Netatmo Energy (thermostat/valves)?

### ISSUE THREAD

**Issue by giuliomagnifico** [xperimental/netatmo-exporter#33]
Hi,
I’m using this exporter and it’s great, thanks for it. 

I’m also using the Netatmo smart thermostat and is very useful to integrate both services in one dashboard. 

At the moment I’m using this exporter for the Energy: https://github.com/tipok/netatmo_energy_exporter

but it’s not well made, the queries are using some weird names, it’s hard to identify the correct sensors and call to it, etc… 
well I’ve done anyway but -since I know nothing of Golang- I was wondering if you can integrate this code into your exporter. I don’t think it should be complicated and this would result in one exporter for all the Netatmo service. 

Obviously if you don’t have time, doesn’t matter. 

Thanks anyway :)

**Comment by xperimental** [comment]
Good idea, I do not have any of these thermostats to test though, so this will be purely based on the documentation...

I think I'll tackle this after the authentication change is released.

**Comment by giuliomagnifico** [comment]
I have the thermostat only at the moment but surely there are people who own also the valves. 

Yes you use the code already written and the docs. If you open a beta branch or whatever I can help you, unfortunately I’m not good in Go language. 

Thanks anyway.

**Comment by giuliomagnifico** [comment]
Hi @xperimental just to update this post, now the old netatmo energy exporter doesn't work anymore due to the autenthication changes from Netatmo, but when you generate the `access` and `refresh`token via Netatmo dev web page, you can also specify to read the thermostat API

![Screenshot 2023-07-14 at 09 26 34](https://github.com/xperimental/netatmo-exporter/assets/1975176/2438b36c-a62d-4fde-a3df-b3ace86a3536)

So I think you have only the add to your exporter the parameters to read the thermostat and valves and it should work.

If you want I can test it, if you make something like a new branch I will clone into it and try to test it.

Thanks anyway

**Comment by xperimental** [comment]
Yes, reading data other than weather will require new scopes.

I have not looked at the API docs for this yet, it's probably also a different endpoint.

We can have a look, once the authentication situation is sorted :slightly_smiling_face:

**Comment by giuliomagnifico** [comment]
@xperimental thanks for the reply, I have no idea of Go language but if there are only some new scopes and a new endpoint I can **try** to modify your exporter adding the new scopes/endopoint and compiling it.

If you have a link to the documentation some little tricks let me know, anyway we have some months before we will use again the thermostat fortunately 😂 (I live in Italy: 45°C now)

**Comment by DEEJCUK** [comment]
https://github.com/tipok/netatmo_energy_exporter
this project reads thermostat and TRV data but its stopped working as it seem the auth flow changed again, any chance your project can implement the thermostat readings ? it would be super nice to get the data from one exporter

**Comment by xperimental** [comment]
Thanks @DEEJCUK . The project was already mentioned in this thread.

I know it's been well over a year, but I can see that other data is the most requested feature for the exporter. I hope that one of the next days will be a "private open-source projects" day again.

**Comment by mighq** [comment]
I am willing to do some work on this. If you have idea about how things should be structured, or already have some code, please share it in the branch. I will post PR, when I have something.

IMO this project is better prepared for adding energy support than `tipok` one. So thanks for good work so far.

**Comment by xperimental** [comment]
Hey @mighq . Thanks for the offer 👍 

I did not have enough time to progress on this issue over the last months. The main issues I see for this feature:

- identify which metrics make sense to export
- check how to integrate this into the authorization flow (make it selectable which data you want)
- either integrate the new metrics into the existing collector or split the logic into two separate collectors for each category

**Comment by xperimental** [comment]
Good news, I actually started looking into this now. It has been mostly looking at the documentation for now and thinking about how to structure this in the exporter, so there's not much concrete to look at yet. But it's a start.

**Comment by giuliomagnifico** [comment]
Hello, in the end I'm using I forked and modified the original repo from [tipok](https://github.com/tipok/netatmo_energy_exporter), now it works perfectly with the refresh token. The original version has many problems and stops working due to authentication errors at every reboot. With my modification, it always works.

If useful, it's here: https://github.com/giuliomagnifico/netatmo_energy_exporter

I have no problems using two exporters. Yes, obviously having everything in one would be more convenient, but I tried to merge both and always had issues. At least this way, everything works.

**Comment by DEEJCUK** [comment]
on the Tipok readme is the instructions on how to run it in docker, can you do the same with this fork

**Comment by DEEJCUK** [comment]
i ran it through ChatGPT and asked it to give me the instructions on how to build it as docker, then using docker compose so you can use it for inspiration

About half way down i asked 

`knowing what you now know, give me step by step instructions to build and run it as a docker container`

and it gave some output.. obviously it may need some tweaking 

https://chatgpt.com/share/67f105b0-43d0-800b-9910-036bd0ee84c4

**Comment by giuliomagnifico** [comment]
@DEEJCUK I don't have a Docker instance to compile and test it. This is the issue I mentioned some month ago. 🙁 I'm sorry

Maybe @xperimental can build the Spcker image, at least for now it's a workaround!

PS: I'm not sure why ChatGPT advises adding the token with JSON formatting. It seems to work fine using plain text without any code formatting. Maybe is working also without the formatting, I haven't tested this

**Comment by giuliomagnifico** [comment]
@daviddanko Yes, I know. I already tried and built it (with Podman on macOS), but I can’t test it. All my Prometheus setup is on “bare metal/compiled” Raspberry PI 5.

If you want to try it, the image is here: https://hub.docker.com/repository/docker/giuliomagnifico/netatmo_energy_exporter/general. However I have no idea if it’s working or what I’ve done… it’s the first time I’ve built one… 😄

**Comment by giuliomagnifico** [comment]
@daviddanko I'm sorry I can't help you, since I'm not using Docker, it's been working for a week and more without issues, even after reboots.

But did you create the `netatmo_token.json` file with the refresh token inside before starting the service? It needs it at the first launch only.

**Comment by giuliomagnifico** [comment]
@daviddanko I think you are doing something wrong, also because the image is always the old one and the executable path is repeated, I can't try but I'll tell you how ChatGPT told me to do it:

```
git clone https://github.com/giuliomagnifico/netatmo_energy_exporter.git
cd netatmo_energy_exporter 
```

Create a Dockerfile in the repository with the following content:

```
FROM golang:1.21-alpine AS builder

WORKDIR /app
COPY . .

RUN go build -o netatmo_energy_exporter

FROM alpine:latest
WORKDIR /app
COPY --from=builder /app/netatmo_energy_exporter /app/netatmo_energy_exporter
COPY netatmo_token.json /app/netatmo_token.json

EXPOSE 2112

ENV NETATMO_TOKEN_FILE=/app/netatmo_token.json

ENTRYPOINT ["/app/netatmo_energy_exporter"]
```
Then build the image:

`docker build -t netatmo-energy-exporter .`

And run it:

```
docker run -d \
  -p 2112:2112 \
  -v /local/path/netatmo_token.json:/app/netatmo_token.json \
  -e NETATMO_CLIENT_ID=xxx \
  -e NETATMO_CLIENT_SECRET=yyy \
  -e NETATMO_TOKEN_FILE=/app/netatmo_token.json \
  netatmo-energy-exporter
```

⚠️ remember to create the `netatmo_token.json` file with the refresh token inside. 

I hope this will work... try it at least.

**Comment by xperimental** [comment]
@giuliomagnifico Looking at the Dockerfile I have some suggestions:

- don't put the application and the configuration into the same directory. That way the configuration directory can be writable while the application directory is not. I suggest to put the binary into one of the "recommended places" for binaries, for example `/bin/`.
- don't run the application as root
- do not put the token file into the container. Instead define the directory as a volume.

Essentially you should be able to mostly re-use the `netatmo-exporter` Dockerfile except for the filenames, port and my name 🙂 

(I have been travelling the whole week and I just realized that it already has been three weeks since my last message ... time flies)

**Comment by xperimental** [comment]
Another update: I did have some time to continue with this effort today, because I was on a train. I am at the point where I can test the data retrieval, but because I don't have any "energy sensors" I don't get any meaningful data.

Would someone be able to send me some raw data from their account (for example per email), so I can have a look? I'm interested in both the `homesdata` as well as the `homestatus` ([NetAtmo documentation](https://dev.netatmo.com/apidocumentation/energy)). There is a "try this out" function in the documentation that you use to get your data.

**Comment by xperimental** [comment]
> I could send you, however, it needs the home_id and I can't seem to find it.

For `homesdata` it is an optional parameter. If you leave it empty, then it will return a list of all your homes. You'll find the ID of each one in the response then and can use that information for the `homestatus` request which is specific for each home.

**Comment by giuliomagnifico** [comment]
@xperimental I can send you my home ID, room details, etc... I can directly send the data from my current exporter's metrics. Just provide an email address to send it to :)

Don't you have the Netatmo thermostat? This is why you need the data? (But I don't have valves...)

**Comment by xperimental** [comment]
> I can send you my home ID, room details, etc... I can directly send the data from my current exporter's metrics. Just provide an email address to send it to :)

Thanks. I'm looking for the raw data from the API not the final metrics created by the other exporter (although that would be nice as well to have a look at what it currently looks like). You can find an address in any commit...

> Don't you have the Netatmo thermostat?

I don't have any of the "newer Netatmo products", only the simple "temperature sensors" for inside and outside. I don't even know if rain / wind works properly, sadly.

**Comment by xperimental** [comment]
It is mentioned in every git commit 🙂

**Comment by xperimental** [comment]
@daviddanko Got your data (thanks!), but did not have any time to take a look at this again so far. Long weekend now ... maybe I'll have some progress on this project as well 🤞

**Comment by giuliomagnifico** [comment]
Hi, now that it’s getting cold and I’m using the thermostat again, I’ve added the thermostat metrics (temperature, heating status, and target temperature) 😁

I’ve opened a PR: https://github.com/xperimental/netatmo-exporter/pull/73

If anyone else wants to try it, I just compiled it, no Docker since I don’t use it on my grafana/promethues server...

Well obviously you need to generate a new token with the "read thermostat" flag!
